state/indexer/sink/psql/psql.go

```diff
@@ -0,0 +1,197 @@
+package psql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	proto "github.com/gogo/protobuf/proto"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/pubsub/query"
+	"github.com/tendermint/tendermint/state/indexer"
+	"github.com/tendermint/tendermint/types"
+)
+
+var _ indexer.EventSink = (*EventSink)(nil)
+
+const (
+	TableEventBlock = "block_events"
+	TableEventTx    = "tx_events"
+	TableResultTx   = "tx_results"
+	DriverName      = "postgres"
+)
+
+// EventSink is an indexer backend providing the tx/block index services.
+type EventSink struct {
+	store   *sql.DB
+	chainID string
+}
+
+func NewEventSink(connStr string, chainID string) (indexer.EventSink, *sql.DB, error) {
+	db, err := sql.Open(DriverName, connStr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &EventSink{
+		store:   db,
+		chainID: chainID,
+	}, db, nil
+}
+
+func (es *EventSink) Type() indexer.EventSinkType {
+	return indexer.PSQL
+}
+
+func (es *EventSink) IndexBlockEvents(h types.EventDataNewBlockHeader) error {
+	sqlStmt := sq.
+		Insert(TableEventBlock).
+		Columns("key", "value", "height", "type", "created_at", "chain_id").
+		PlaceholderFormat(sq.Dollar)
+
+	ts := time.Now()
+	// index the reserved block height index
+	sqlStmt = sqlStmt.
+		Values(types.BlockHeightKey, fmt.Sprint(h.Header.Height), h.Header.Height, "", ts, es.chainID)
+
+	// index begin_block events
+	sqlStmt, err := indexBlockEvents(
+		sqlStmt, h.ResultBeginBlock.Events, types.EventTypeBeginBlock, h.Header.Height, ts, es.chainID)
+	if err != nil {
+		return err
+	}
+
+	// index end_block events
+	sqlStmt, err = indexBlockEvents(
+		sqlStmt, h.ResultEndBlock.Events, types.EventTypeEndBlock, h.Header.Height, ts, es.chainID)
+	if err != nil {
+		return err
+	}
+
+	_, err = sqlStmt.RunWith(es.store).Exec()
+	return err
+}
+
+func (es *EventSink) IndexTxEvents(txr []*abci.TxResult) error {
+	// index the tx result
+	var txid uint32
+	sqlStmtTxResult := sq.
+		Insert(TableResultTx).
+		Columns("tx_result", "created_at").
+		PlaceholderFormat(sq.Dollar).
+		RunWith(es.store).
+		Suffix("RETURNING \"id\"")
```

Comment:
perfer raw strings with backtics rather than escaping quote marks?

Reply:
tried, it didn't work.